### PR TITLE
Optuna Implementation

### PR DESCRIPTION
**Pull request by emmatekulova**
Ahoj,

pridala som súbor `hyperparam.py`, v ktorom je vytvorená pipeline pre tuning hyperparametrov.

---

### Evaluate metric

* Evaluačná metrika je ako `Callable`.
* Uvedomila som si, že spracovávať `dict` je zbytočné, keďže potrebujem iba jednu `float` hodnotu.
* Kombinované metriky môžu závisieť od toho, čo usert chce optimalizovať - takže je to jeho zodpovednosť.

---

### Ouputs folders

Pipeline automaticky vytvorí tri folders:

1. `lightning_logs` - výstupy z PyTorch Lightning logging
2. `checkpoints` - modely uložené cez `ModelCheckpoint` callback
3. `output` - obsahuje `.pt` modely a `.json` súbory 

---

### Plans

Rada by som ešte pridala support pre ďalšie typy parametrov - nielen `categorical`,.

Na to sa ešte potrebujem pozrieť, ako to čo najzrozumiteľnejšie vyriešiť.

---

### Notebooks

Aktuálna verzia je plne kompatibilná s notebookom:

* Stačí ju spustiť,
* definovať search space pre hyperparametre,

Poprosím o review, Ďakujem !

### CODE REVIEW

File: src/seglight/hyperparam.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+
+class OptunaLightningTuner:
+    def __init__(
+        self,
+        model_builder: Callable,
+        model_class: type[LightningModule],
+        loss_fn: Callable,
+        datamodule: TrainTestDataModule,
+        param_search_space: dict,  # dict: {param_name: list_of_values}
```
Comment (jaroslavknotek) [inline src/seglight/hyperparam.py:59]: Pokud tento kód je skutečně externí knihovna instalovaná přes pip, tak jak co musí udělat user tohoto kódu, aby si přečetl tento komentář? Tenhle skript je uložený někde v cestě jako je `/home/foo/.config/conda/x.y.z/site-packages/libs/seglights/abx/hyperparam.py`. Komentáře nejsou pro usery, ale pro vývojáře.

**Review by jaroslavknotek — CHANGES_REQUESTED** [review]
Jedna drobná změna a jedna architektonická a jsme tam.
